src/sbvr-api/sbvr-utils.coffee

```diff
@@ -353,18 +356,65 @@ getHooks = do ->
 	_getHooks = (request) ->
 		if request.resourceName?
 			resourceName = resolveSynonym(request)
-		getMethodHooks(request.method, request.vocabulary, resourceName)
+		instantiateHooks(
+			getMethodHooks(
+				request.method
+				request.vocabulary
+				resourceName
+			)
+		)
 	_getHooks.clear = -> getMethodHooks.clear()
 	return _getHooks
 
-runHook = (hookName, args) ->
+class Hook
+	constructor: (opts) ->
+		@hookFn = opts.HOOK
+		@rollbackFn = opts.ROLLBACK
+		@rollbackArgs = []
+		@executed = false
+
+	setArgument: (arg) ->
+		if @rollbackArgs.length < @rollbackFn.length
+			@rollbackArgs.push(arg)
+		else
+			throw new Error('Too many arguments supplied to rollback function')
+
+	run: (args...) ->
+		Promise.try =>
+			@hookFn(args...)
+		.tap =>
+			if @rollbackArgs.length != @rollbackFn.length
+				throw new Error('registerRollback not supplied enough arguments')
+			@executed = true
+		# If we detect that some rollbacks are registered but the rollback function is not saturated, we add the missing number of arguments as null.
+		# The ROLLBACK function should take care of checking and performing the correct number of rollback actions.
+		.tapCatch =>
+			if not _.isEmpty(@rollbackArgs)
+				missingArgs = @rollbackFn.length - @rollbackArgs.length
+				for i in _.range(missingArgs)
+					@rollbackArgs.push(null)
+				@rollback()
+
+	registerRollback: (args...) ->
+		for arg in args
```

Comment:
I think this would be better as:
```coffee
registerRollback: (@rollbackArgs...) ->
	if @rollbackArgs.length > @rollbackFn.length
		throw new Error('Too many arguments supplied to rollback function')
```
and remove `setArgument` altogether.  Although that said the length check is a bit awkward altogether, eg default values don't count towards the function length but are still valid to supply:
```js
foo = (a = 'b') => {}
foo.length // 0
```

Reply:
The idea of the `setArgument` is to make hooks with multiple side-effects less awkward. So we can re-use the supplied rollback action as partial cancellation of side-effects if something goes wrong in the hook.
In general rollback functions should not define default arguments, but check and only perform the cancellation actions for the supplied arguments.

Reply:
Oh, that is definitely not how I was envisioning rollback, how do you handle the following case
```coffee
HOOK: ({ ..., registerRollback }) ->
	Promise.all([
		createExternalResource(1)
		.then (id1) ->
			registerRollback(id1)

		createExternalResource(2)
		.then (id2) ->
			registerRollback(id2)
	])
```
where the rollbacks can be registered in different orders, or
```coffee
HOOK: ({ ..., registerRollback }) ->
	Promise.try ->
		if test()
			createExternalResource(1)
			.then (id1) ->
				registerRollback(id1)
	.then ->
		# Additional logic
		createExternalResource(2)
		.then (id2) ->
			registerRollback(id2)
		.catch (err) ->
			# No need to delete resource 1 here
```
where one of the early creations might be skipped?
I was envisioning something much more like
```coffee
HOOK: ({ ..., registerRollback }) ->
	Promise.all([
		createExternalResource(1)
		.then (id1) ->
			registerRollback ->
				deleteExternalResource(id1)

		createExternalResource(2)
		.then (id2) ->
			registerRollback ->
				deleteExternalResource(id2)
	])
```
so that we just past a selection of functions to be run on rollback rather than relying on a fragile order of adding arguments